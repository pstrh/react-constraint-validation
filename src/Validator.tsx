
export type Messages = {
    [key: string]: string
}

const defaultMessages: Messages = {
    "required": "{name} is required",
    "min": "{name} must be greater than {minValue}",
    "max": "{name} must be less than {maxValue}",
    "minLength": "{name} must have at least {minLength} characters",
    "maxLength": "{name} must have less than {maxLength} characters",
    "pattern": "{name} does not match the pattern {pattern}",
    "number": "{name} must be a number",
    "email": "{name} is not a valid email"
};

type InterpolationPatternFn =  (param: string) => string;

const defaultInterpolationPatternFn: InterpolationPatternFn = (param: string): string => {
    return `{${param}}`;
};

let configuredMessages: Messages = defaultMessages;
let interpolationPatternFn: InterpolationPatternFn = defaultInterpolationPatternFn;

export function init(overrideMessages: Messages) {
    configuredMessages = {...defaultMessages, ...overrideMessages};
}

export function setInterpolationPatternFn(overrideInterpolationPatternFn: InterpolationPatternFn) {
    interpolationPatternFn = overrideInterpolationPatternFn;
}

export function getErrorMessage(validatorKey: string, name: string | undefined) {
    if (!name) {
        return configuredMessages[validatorKey];
    } else {
        const messageWithName: string | undefined = configuredMessages[validatorKey + "." + name];
        if (messageWithName) {
            return messageWithName;
        } else {
            return configuredMessages[validatorKey];
        }
    }
}

type InterpolationParams = {
    [key: string]: any
}

export function interpolate(message: string, params: InterpolationParams): string {
    let finalMessage = message;
    Object.keys(params).forEach(function(key: string) {
        const value: any = params[key];
        if (value) {
            finalMessage = finalMessage.replace(interpolationPatternFn(key), value);
        }
    });
    return finalMessage;
}

export function resolveErrorMessage(validatorKey: string, name: string | undefined, interpolationParams: InterpolationParams) {
    return interpolate(getErrorMessage(validatorKey, name), interpolationParams);
}

export function required(name: string | undefined, value: any): string | undefined {
    if (!value) {
        return resolveErrorMessage("required", name, {name, value});
    }
    return undefined;
}

export function min(name: string | undefined, value: string, minValue: number): string | undefined {
    if (!value) {
        return undefined;
    }
    const valueAsNumber = parseFloat(value);
    if (valueAsNumber && valueAsNumber < minValue) {
        return resolveErrorMessage("min", name, {name, value, minValue});
    }
    return undefined;
}

export function max(name: string | undefined, value: string, maxValue: number): string | undefined {
    if (!value) {
        return undefined;
    }
    const valueAsNumber = parseFloat(value);
    if (valueAsNumber && valueAsNumber > maxValue) {
        return resolveErrorMessage("max", name, {name, value, maxValue});
    }
    return undefined;
}

export function minLength(name: string | undefined, value: string, minLength: number): string | undefined {
    if (!value) {
        return undefined;
    }
    if (value.length < minLength) {
        return resolveErrorMessage("minLength", name, {name, value, minLength});
    }
    return undefined;
}

export function maxLength(name: string | undefined, value: string, maxLength: number): string | undefined {
    if (!value) {
        return undefined;
    }
    if (value.length > maxLength) {
        return resolveErrorMessage("maxLength", name, {name, value, maxLength});
    }
    return undefined;
}

export function pattern(name: string | undefined, value: string, pattern: string): string | undefined {
    if (!value) {
        return undefined;
    }
    const regEx = new RegExp(pattern);
    if (!regEx.test(value)) {
        return resolveErrorMessage("pattern", name, {name, value, pattern});
    }
    return undefined;
}

export function email(name: string | undefined, value: string): string | undefined {
    if (!value) {
        return undefined;
    }
    // email regex taken from https://stackoverflow.com/questions/46155/how-to-validate-an-email-address-in-javascript
    if (maxLength(name, value, 254) || pattern(name, value, "^(([^<>()\\[\\]\\.,;:\\s@\\\"]+(\\.[^<>()\\[\\]\\.,;:\\s@\\\"]+)*)|(\\\".+\\\"))@(([^<>()[\\]\\.,;:\\s@\\\"]+\\.)+[^<>()[\\]\\.,;:\\s@\\\"]{2,})$")) {
        return resolveErrorMessage("email", name, {name, value});
    }
    return undefined;
}

export function number(name: string | undefined, value: string): string | undefined {
    if (!value) {
        return undefined;
    }
    const valueAsNumber = parseFloat(value);
    if (!valueAsNumber) {
        return resolveErrorMessage("number", name, {name, value});
    }
    return undefined;
}

