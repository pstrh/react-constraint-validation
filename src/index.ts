export * from "./Validator"
export * from "./withValidator";