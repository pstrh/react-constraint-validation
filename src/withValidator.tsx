import * as React from "react";

export type ValidationFunction = (name: string | undefined, value:any) => string | undefined;
export type ValidationFunctionWithParam = (name: string | undefined, value:any, param: any) => string | undefined;

type Validators = {[key:string]: ValidationFunction | ValidationFunctionWithParam};
type DefaultValidators = {[key:string]: ValidationFunction};

type ValidationProp<T> = {
    [P in keyof T]: any
}

interface FieldWithValidation {
    name?: string,
    validate?: (name: string, value:any) => string | undefined
}

interface ComponentDecorator<T> {
    <P extends FieldWithValidation>(component: React.ComponentType<P>): React.ComponentClass<P & Partial<ValidationProp<T>>>;
}

// TODO passThroughProps
// TODO validate props configuration (adaption for redux-form etc.)
// TODO add hoist static
export function withValidator<T extends Validators>(validators: T, defaultValidators?: DefaultValidators): ComponentDecorator<T> {

    return function createValidationWrapperHOC<P extends FieldWithValidation>(WrappedComponent: React.ComponentType<P>): React.ComponentClass<P & Partial<ValidationProp<T>>> {

        class ValidationWrapperHOC extends React.Component<P & Partial<ValidationProp<T>>> {

            constructor(props: P & Partial<ValidationProp<T>>) {
                super(props);
                this.validate = this.validate.bind(this);
            }

            validate(value: any): string | undefined {
                let error = this.applyDefaultValidators(defaultValidators, this.props.name, value);
                if (!error) {
                    error = this.applyValidators(validators, this.props.name, value);
                }
                return error;
            }

            applyValidators(validators: Validators, name: string | undefined, value: string): string | undefined {
                if (!validators) {
                    return undefined;
                }
                const props = this.props;
                let firstError: string | undefined = undefined;
                Object.keys(validators).some(function(key: string) {
                    const validatorParam = props[key];
                    if (validatorParam) {
                        if (typeof validatorParam === "boolean") {
                            const validate = validators[key] as ValidationFunction;
                            firstError = validate(name, value);
                        } else {
                            const validate = validators[key] as ValidationFunctionWithParam;
                            firstError = validate(name, value, validatorParam);
                        }
                    }
                    return firstError != undefined;
                });
                return firstError;
            }

            applyDefaultValidators(defaultValidators: undefined | DefaultValidators, name: string | undefined, value: string): string | undefined {
                if (!defaultValidators) {
                    return undefined;
                }
                let firstError: string | undefined = undefined;
                Object.keys(defaultValidators).some(function(key: string) {
                    const validate = defaultValidators[key] as ValidationFunction;
                    firstError = validate(name, value);
                    return firstError != undefined;
                });
                return firstError;
            }

            removeValidatorProps(props: Readonly<P & Partial<ValidationProp<T>>>): P {
                const clonedProps = Object.assign({}, props);
                Object.keys(validators).forEach(function(key: string) {
                    //@ts-ignore
                    delete clonedProps[key];
                });
                //@ts-ignore
                return clonedProps as P;
            }

            render() {
                const originalProps = this.removeValidatorProps(this.props);

                return (
                    <WrappedComponent {...originalProps} validate={this.validate}/>
                );
            }
        };

        return ValidationWrapperHOC;
    }
}


