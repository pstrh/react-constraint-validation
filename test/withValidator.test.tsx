import {
    number,
    max, maxLength,
    min,
    minLength,
    required, pattern, email
} from "../src/Validator";
import * as React from "react";
import * as renderer from "react-test-renderer";
import {withValidator} from "../src/withValidator";

interface FieldProps {
    name?: string,
    value?: string,
    validate?: (value:any) => string | undefined
}

const Field: React.SFC<FieldProps> = (props) => {
    const validationResult: string | undefined = props.validate ? props.validate(props.value) : undefined;
    return <>{"validateTest: " + (validationResult ? validationResult : "valid")}</>;
};

const NumberField = withValidator({required, min, max },{number})(Field);

const TextField = withValidator({required, minLength, maxLength, pattern})(Field);

const EmailField = withValidator({required},{email})(Field);

const NumberFieldTest: React.SFC = () => {
    return (
        <>
            <NumberField name="test"/>
            <NumberField name="test" required/>
            <NumberField name="test" value="a" required/>
            <NumberField name="test" value="1" required min={2}/>
            <NumberField name="test" value="5" required min={2} max={4}/>
            <NumberField name="test" value="3" required min={2} max={4}/>
        </>
    );
};

test("NumberField test", () => {
    const tree = renderer.create(<NumberFieldTest/>).toJSON();
    expect(tree).toMatchSnapshot();
});

const TextFieldTest: React.SFC = () => {
    return (
        <>
            <TextField name="test"/>
            <TextField name="test" required/>
            <TextField name="test" value="abcdef" required/>
            <TextField name="test" value="ab" required minLength={3}/>
            <TextField name="test" value="abcdef" required minLength={3} maxLength={5}/>
            <TextField name="test" value="abcd" required minLength={3} maxLength={5}/>
            <TextField name="test" value="abcd123" required pattern={"^[a-z0-9_\\-]+$"}/>
            <TextField name="test" value="aBcd123" required pattern={"^[a-z0-9_\\-]+$"}/>
        </>
    );
};

test("TextField test", () => {
    const tree = renderer.create(<TextFieldTest/>).toJSON();
    expect(tree).toMatchSnapshot();
});

const EmailFieldTest: React.SFC = () => {
    return (
        <>
            <EmailField name="test"/>
            <EmailField name="test" required/>
            <EmailField name="test" value="test" required/>
            <EmailField name="test" value="test@test" required/>
            <EmailField name="test" value="test@test.com" required/>
        </>
    );
};

test("EmailField test", () => {
    const tree = renderer.create(<EmailFieldTest/>).toJSON();
    expect(tree).toMatchSnapshot();
});