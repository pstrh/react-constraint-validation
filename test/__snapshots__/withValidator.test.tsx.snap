// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailField test 1`] = `
Array [
  "validateTest: valid",
  "validateTest: test is required",
  "validateTest: test is not a valid email",
  "validateTest: test is not a valid email",
  "validateTest: valid",
]
`;

exports[`NumberField test 1`] = `
Array [
  "validateTest: valid",
  "validateTest: test is required",
  "validateTest: test must be a number",
  "validateTest: test must be greater than 2",
  "validateTest: test must be less than 4",
  "validateTest: valid",
]
`;

exports[`TextField test 1`] = `
Array [
  "validateTest: valid",
  "validateTest: test is required",
  "validateTest: valid",
  "validateTest: test must have at least 3 characters",
  "validateTest: test must have less than 5 characters",
  "validateTest: valid",
  "validateTest: valid",
  "validateTest: test does not match the pattern ^[a-z0-9_\\\\-]+$",
]
`;
